external/fv3fit/tests/keras/test_halos.py

```diff
@@ -0,0 +1,159 @@
+    "nx, ny, nz, n_tile, n_halo", [pytest.param(8, 8, 12, 1, 3, id="typical")]
+)
+def test_append_halos_raises_with_one_tile(
+    nx: int, ny: int, nz: int, n_tile: int, n_halo: int
+):
+    ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile)
+    with pytest.raises(ValueError):
+        append_halos(ds, n_halo=n_halo)
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_tile", [pytest.param(8, 8, 12, 6, id="typical")]
+)
+def test_append_halos_no_halos_is_unchanged(nx: int, ny: int, nz: int, n_tile: int):
+    ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile)
+    result: xr.Dataset = append_halos(ds, n_halo=0)
+    xr.testing.assert_identical(result, ds)
+
+
+class ChainedDummyComm(fv3gfs.util.testing.DummyComm):
+    """
+    Dummy comm that calls a callback just before recv, allowing us to call all sends
+    before recvs.
+
+    Can be used to mock MPI behavior in a function that has a single send-recv pair.
+    """
+
+    def __init__(self, rank, total_ranks, buffer_dict):
+        super().__init__(rank=rank, total_ranks=total_ranks, buffer_dict=buffer_dict)
+        self._callback = lambda: None
+
+    def set_callback(self, callback):
+        self._callback = callback
+
+    def Recv(self, *args, **kwargs):
+        self._callback()
+        self._callback = lambda: None
+        super().Recv(*args, **kwargs)
+
+
+def _get_callback(output_datasets, i: int, rank_dataset, n_halo, comm):
+    def callback():
+        output_datasets[i] = _append_halos_using_mpi(
+            ds=rank_dataset, n_halo=n_halo, comm=comm
+        )
+
+    return callback
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_halo",
+    [pytest.param(8, 8, 12, 3, id="typical"), pytest.param(8, 8, 12, 1, id="one_halo")],
+)
+def test_with_and_without_mpi_give_same_result(nx: int, ny: int, nz: int, n_halo: int):
+    n_tile = 6
+    buffer_dict: Dict[Any, Any] = {}
+    comms: List[ChainedDummyComm] = []
+    for rank in range(n_tile):
+        comms.append(
+            ChainedDummyComm(rank=rank, total_ranks=n_tile, buffer_dict=buffer_dict)
+        )
+    full_ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile).drop(
+        ["scalar", "scalar_single_precision"]
+    )
+    non_mpi_result: xr.Dataset = append_halos(full_ds, n_halo=n_halo)
+    rank_datasets = [full_ds.isel(tile=i) for i in range(n_tile)]
+
+    output_datasets = [None for _ in range(n_tile)]
+    for i in range(1, n_tile):
+        comms[i - 1].set_callback(
+            _get_callback(
+                output_datasets,
+                i=i,
+                rank_dataset=rank_datasets[i],
+                n_halo=n_halo,
+                comm=comms[i],
+            )
+        )
+    output_datasets[0] = _append_halos_using_mpi(
+        ds=rank_datasets[0], n_halo=n_halo, comm=comms[0]
+    )
+    for i in range(n_tile):
+        check_result(result=output_datasets[i], n_halo=n_halo, nx=nx, ny=ny, nz=nz)
+        xr.testing.assert_identical(non_mpi_result.isel(tile=i), output_datasets[i])
```

Comment:
Could just `rank_datasets[i]` instead of non_mpi_result

Reply:
I tried then realized I actually can't. I modified the code above to compute rank_datasets before it computes non_mpi_result to make this clearer. rank_datasets is input data, while non_mpi_result has halos appended.